qiskit_ibm_runtime/estimator.py

```diff
@@ -146,8 +147,19 @@ def run(
             Submitted job.
 
         """
+        # Coerce pubs to validate input pub-like arguments
         coerced_pubs = [EstimatorPub.coerce(pub, precision) for pub in pubs]
-        return self._run(coerced_pubs)  # type: ignore[arg-type]
+        # Unpack pubs back to Pub-like tuples for serialization
+        unpacked = [
+            (
+                pub.circuit,
+                np.asarray(pub.observables),
+                pub.parameter_values.as_array(pub.circuit.parameters),
+                pub.precision,
+            )
+            for pub in coerced_pubs
+        ]
```

Comment:
Why not do this in the `RuntimeEncoder` to keep the code here clean?

Reply:
It's not round-trip able if you put it there, and the point of this was to avoid having special encoding just for the program inputs that you know are pubs.

Reply:
This also allows us to later add proper round-trippable JSON encoding/decoding into the encoder which I think we want eventually once everything else is working properly

Reply:
I don't understand why not. `RuntimeEncoder` can convert an `EstimatorPub` to a tuple like this. And `RuntimeDecoder` continues to do nothing (since it wouldn't know if a tuple is a pub). Then server gets a list of tuples under the `pubs` keyword and knows they are pubs and can convert them back to pubs.

Reply:
Ok, i added unpacking to RuntimeEncoder and removed them from the deceoder